File: src/main/tasks/InstallQueueService.ts
import fs from 'node:fs'
import { randomUUID } from 'node:crypto'
import { ipcChannels, InstallQueueEventSchema, InstallQueueItemSchema, type InstallQueueEvent, type InstallQueueItem } from '@shared/contracts'
import { deriveLabelFromFilePath } from '@shared/utils/display'
import { DiagnosticsService } from '@main/services/diagnostics/DiagnosticsService'
import { AdbService } from '@main/services/adb/AdbService'
import { SettingsStore } from '@main/storage/SettingsStore'
import { WsaService } from '@main/services/wsa/WsaService'

type Publisher = (channel: string, payload: unknown) => void

export class InstallQueueService {
  private readonly items: InstallQueueItem[] = []
  private processing = false
  private publisher?: Publisher

  constructor(
    private readonly adbService: AdbService,
    private readonly wsaService: WsaService,
    private readonly settingsStore: SettingsStore,
    private readonly diagnostics: DiagnosticsService
  ) {}

  setPublisher(publisher: Publisher): void {
    this.publisher = publisher
  }

  getItems(): InstallQueueItem[] {
    return [...this.items]
  }

  async enqueue(paths: string[]): Promise<InstallQueueItem[]> {
    const created: InstallQueueItem[] = []

    for (const filePath of paths) {
      const fileName = filePath.split(/[\\/]/).at(-1) ?? filePath
      const stats = fs.existsSync(filePath) ? fs.statSync(filePath) : undefined
      const isApk = filePath.toLowerCase().endsWith('.apk')
      const item = InstallQueueItemSchema.parse({
        id: randomUUID(),
        filePath,
        fileName,
        displayName: deriveLabelFromFilePath(filePath),
        sizeBytes: stats?.size ?? 0,
        state: isApk && stats?.isFile() ? 'queued' : 'rejected',
        progress: 0,
        error: isApk && stats?.isFile() ? undefined : 'Only existing .apk files are supported in v1.',
        createdAt: new Date().toISOString(),
        updatedAt: new Date().toISOString()
      })

      this.items.unshift(item)
      created.push(item)
    }

    this.emitSnapshot()
    void this.processQueue()
    return created
  }

  private emitSnapshot(): void {
    const event = InstallQueueEventSchema.parse({
      type: 'snapshot',
      items: this.items
    })
    this.publisher?.(ipcChannels.queueEvent, event)
  }

  private updateItem(id: string, patch: Partial<InstallQueueItem>): void {
    const index = this.items.findIndex((item) => item.id === id)
    if (index === -1) {
      return
    }

    this.items[index] = InstallQueueItemSchema.parse({
      ...this.items[index],
      ...patch,
      updatedAt: new Date().toISOString()
    })

    const event: InstallQueueEvent = InstallQueueEventSchema.parse({
      type: 'updated',
      items: this.items
    })
    this.publisher?.(ipcChannels.queueEvent, event)
  }

  private async processQueue(): Promise<void> {
    if (this.processing) {
      return
    }

    this.processing = true
    try {
      while (true) {
        const nextItem = this.items.find((item) => item.state === 'queued')
        if (!nextItem) {
          break
        }

        await this.processItem(nextItem)
      }
    } finally {
      this.processing = false
    }
  }

  private async processItem(item: InstallQueueItem): Promise<void> {
    this.updateItem(item.id, { state: 'connecting', progress: 10 })
    this.diagnostics.log('info', 'queue', `Preparing install for ${item.fileName}`)

    try {
      await this.adbService.ensureServer()
      let connection = await this.wsaService.connectToBestEndpoint(this.settingsStore.get().manualEndpoint)

      if (!connection.connected) {
        await this.wsaService.wake()
        await new Promise((resolve) => setTimeout(resolve, 2_500))
        connection = await this.wsaService.connectToBestEndpoint(this.settingsStore.get().manualEndpoint)
      }

      if (!connection.connected) {
        throw new Error(connection.message)
      }

      this.updateItem(item.id, { state: 'installing', progress: 45 })
      const beforePackages = new Set(await this.adbService.listUserPackages())
      const installResult = await this.adbService.installApk(item.filePath)

      if (!installResult.success) {
        throw new Error(installResult.message)
      }

      const afterPackages = await this.adbService.listUserPackages()
      const newPackage = afterPackages.find((packageName) => !beforePackages.has(packageName))

      this.updateItem(item.id, {
        state: 'installed',
        progress: 100,
        packageName: newPackage
      })

      if (newPackage) {
        this.settingsStore.recordInstall({
          packageName: newPackage,
          label: item.displayName,
          fileName: item.fileName,
          installedAt: new Date().toISOString()
        })
      }

      this.diagnostics.log('success', 'queue', `Installed ${item.fileName}`, newPackage ? `Package: ${newPackage}` : undefined)
    } catch (error) {
      this.updateItem(item.id, {
        state: 'failed',
        progress: 100,
        error: error instanceof Error ? error.message : String(error)
      })
      this.diagnostics.log('error', 'queue', `Install failed for ${item.fileName}`, error instanceof Error ? error.message : String(error))
    }
  }
}

File: WsaManager.Core/InstallQueueService.cs
namespace WsaManager.Core;

public sealed class InstallQueueService
{
    private readonly List<InstallQueueItem> items = [];
    private readonly AdbService adbService;
    private readonly WsaService wsaService;
    private readonly SettingsStore settingsStore;
    private readonly DiagnosticsService diagnostics;
    private readonly IApkPackageReader apkPackageReader;
    private readonly ApkIconExtractor iconExtractor = new();
    private bool processing;

    public InstallQueueService(AdbService adbService, WsaService wsaService, SettingsStore settingsStore, DiagnosticsService diagnostics, IApkPackageReader? apkPackageReader = null)
    {
        this.adbService = adbService;
        this.wsaService = wsaService;
        this.settingsStore = settingsStore;
        this.diagnostics = diagnostics;
        this.apkPackageReader = apkPackageReader ?? new ApkPackageReader();
    }

    public event EventHandler<IReadOnlyList<InstallQueueItem>>? QueueChanged;

    public IReadOnlyList<InstallQueueItem> Items => items.ToList();

    public async Task<IReadOnlyList<InstallQueueItem>> EnqueueAsync(IEnumerable<string> paths, CancellationToken cancellationToken = default)
    {
        var created = new List<InstallQueueItem>();
        foreach (var filePath in paths)
        {
            var fileInfo = new FileInfo(filePath);
            var isApk = filePath.EndsWith(".apk", StringComparison.OrdinalIgnoreCase);
            var isValid = isApk && fileInfo.Exists;
            var item = new InstallQueueItem
            {
                Id = Guid.NewGuid().ToString("n"),
                FilePath = filePath,
                FileName = Path.GetFileName(filePath),
                DisplayName = DisplayNames.DeriveLabelFromFilePath(filePath),
                SizeBytes = fileInfo.Exists ? fileInfo.Length : 0,
                State = isValid ? QueueItemState.Queued : QueueItemState.Rejected,
                Progress = 0,
                Error = isValid ? null : "Only existing .apk files are supported in v1.",
                CreatedAt = DateTimeOffset.UtcNow,
                UpdatedAt = DateTimeOffset.UtcNow
            };

            items.Insert(0, item);
            created.Add(item);
        }

        EmitSnapshot();
        await ProcessQueueAsync(cancellationToken);
        return created;
    }

    public async Task ReplaceExistingAsync(string itemId, CancellationToken cancellationToken = default)
    {
        var item = items.FirstOrDefault(item => item.Id == itemId);
        if (item is null || item.State != QueueItemState.AlreadyInstalled)
        {
            return;
        }

        await ProcessItemAsync(item, forceReplaceExisting: true, cancellationToken);
        await ProcessQueueAsync(cancellationToken);
    }

    public async Task SkipAsync(string itemId, CancellationToken cancellationToken = default)
    {
        var item = items.FirstOrDefault(item => item.Id == itemId);
        if (item is null || item.State != QueueItemState.AlreadyInstalled)
        {
            return;
        }

        item.Error = null;
        UpdateItem(item, QueueItemState.Skipped, 100);
        diagnostics.Log(DiagnosticLevel.Info, "queue", $"Skipped reinstall for {item.PackageName ?? item.FileName}");
        await ProcessQueueAsync(cancellationToken);
    }

    private async Task ProcessQueueAsync(CancellationToken cancellationToken)
    {
        if (processing)
        {
            return;
        }

        processing = true;
        try
        {
            while (items.FirstOrDefault(item => item.State == QueueItemState.Queued) is { } nextItem)
            {
                await ProcessItemAsync(nextItem, forceReplaceExisting: false, cancellationToken);
            }
        }
        finally
        {
            processing = false;
        }
    }

    private async Task ProcessItemAsync(InstallQueueItem item, bool forceReplaceExisting, CancellationToken cancellationToken)
    {
        UpdateItem(item, QueueItemState.Connecting, 10);
        diagnostics.Log(DiagnosticLevel.Info, "queue", $"Preparing install for {item.FileName}");

        try
        {
            await adbService.EnsureServerAsync(cancellationToken: cancellationToken);
            var detectedPackageName = apkPackageReader.TryReadPackageName(item.FilePath);
            item.PackageName = detectedPackageName;
            EmitSnapshot();

            var connection = await wsaService.AutoHealConnectionAsync(settingsStore.Get().ManualEndpoint, cancellationToken: cancellationToken);

            if (!connection.Connected)
            {
                throw new InvalidOperationException(connection.Message);
            }

            UpdateItem(item, QueueItemState.Installing, 45);
            var beforePackages = new HashSet<string>(await adbService.ListUserPackagesAsync(cancellationToken: cancellationToken));
            if (!string.IsNullOrWhiteSpace(detectedPackageName) && beforePackages.Contains(detectedPackageName))
            {
                if (!forceReplaceExisting)
                {
                    item.Error = "This app is already installed. Choose Replace app to reinstall it, or Skip to leave it alone.";
                    UpdateItem(item, QueueItemState.AlreadyInstalled, 0);
                    diagnostics.Log(DiagnosticLevel.Warn, "queue", $"{detectedPackageName} is already installed");
                    return;
                }

                var uninstallResult = await adbService.UninstallPackageAsync(detectedPackageName, cancellationToken: cancellationToken);
                if (!uninstallResult.Success)
                {
                    throw new InvalidOperationException($"Could not uninstall existing {detectedPackageName}: {uninstallResult.Message}");
                }

                item.WasUninstalledBeforeInstall = true;
                diagnostics.Log(DiagnosticLevel.Info, "queue", $"Uninstalled existing {detectedPackageName} before reinstall");
            }

            var installResult = await adbService.InstallApkAsync(item.FilePath, cancellationToken: cancellationToken);
            if (!installResult.Success)
            {
                throw new InvalidOperationException(installResult.Message);
            }

            var afterPackages = await adbService.ListUserPackagesAsync(cancellationToken: cancellationToken);
            var newPackage = detectedPackageName ?? afterPackages.FirstOrDefault(packageName => !beforePackages.Contains(packageName));
            item.PackageName = newPackage;
            var iconPath = iconExtractor.TryExtractIcon(item.FilePath, Path.Combine(settingsStore.AppDataDirectory, "icons"), newPackage ?? item.DisplayName);
            settingsStore.RecordInstall(new InstallHistoryEntry
            {
                PackageName = newPackage ?? item.DisplayName,
                Label = item.DisplayName,
                FileName = item.FileName,
                SizeBytes = item.SizeBytes,
                IconPath = iconPath,
                InstalledAt = DateTimeOffset.UtcNow
            });
            UpdateItem(item, QueueItemState.Installed, 100);
            diagnostics.Log(DiagnosticLevel.Success, "queue", $"Installed {item.FileName}", newPackage is not null ? $"Package: {newPackage}" : null);
        }
        catch (Exception ex)
        {
            item.Error = ex.Message;
            UpdateItem(item, QueueItemState.Failed, 100);
            diagnostics.Log(DiagnosticLevel.Error, "queue", $"Install failed for {item.FileName}", ex.Message);
        }
    }

    private void UpdateItem(InstallQueueItem item, QueueItemState state, int progress)
    {
        item.State = state;
        item.Progress = progress;
        item.UpdatedAt = DateTimeOffset.UtcNow;
        EmitSnapshot();
    }

    private void EmitSnapshot() => QueueChanged?.Invoke(this, Items);
}
